### ISSUE TITLE
dfn/reduce-+ NPE, depending on data size

### ISSUE THREAD

**Issue by daslu** [techascent/tech.datatype#17]
```clojure
$ clj -Sdeps '{:deps {techascent/tech.datatype {:mvn/version "4.65"}}}}'
Clojure 1.10.1
user=> (require '[tech.v2.datatype.functional :as dfn])
nil
user=> (-> (range 10) dfn/reduce-+)
Execution error (IllegalArgumentException) at tech.v2.datatype.operation-provider/eval15184$fn (operation_provider.clj:79).
No method in multimethod 'half-dispatch-reduce-op' for dispatch value: :iterable
user=> ;; Seem to be missing tech.v2.datatype. Let us require it.
user=> (require '[tech.v2.datatype :as dtype])
nil
user=> (-> (range 10) dfn/reduce-+)
45
user=> ;; Now it works. Also in the reader version:
user=> (-> (range 10) double-array dfn/reduce-+)
45.0
user=> (-> (range 10) into-array dfn/reduce-+)
45
user=> ;; What about more data?
user=> (-> (range 1000) dfn/reduce-+)
499500
user=> (-> (range 1000) double-array dfn/reduce-+)
499500.0
user=> (-> (range 1000) into-array dfn/reduce-+)
Execution error (NullPointerException) at tech.v2.datatype.binary_op$reify$reify__8062/op (binary_op.clj:462).
null
user=> ;; Now this is confusing.
user=> ;; What about a different container, with an explicit type?
user=> (->> (range 1000) into-array (dtype/make-container :java-array :double) dfn/reduce-+)
Execution error (NullPointerException) at tech.v2.datatype.binary_op$reify$reify__8062/op (binary_op.clj:462).
null
user=> ;; Still NPE.
user=> 

```

**Comment by cnuernber** [comment]
This is a good bug.  There are several shortcomings here in datatype that I need to address.  The reason the make-container pathway failed is that `:double` isn't a primitive datatype (although maybe it should be.  `float64` is.  Regardless, reduce-+ should work with object datatypes and simply requiring the functional namespace should be enough to do the above samples.  Great bug, thanks, working on it.

**Comment by cnuernber** [comment]
Fixed (with test) in 4.67

**Comment by cnuernber** [comment]
@daslu - 

All these examples now not only work but are a bit faster.  Reduce-+ hits a parallelized reduction pathway for commutative operators; that is what was originally causing the bug but recent updates to datatype have made all of that much better.
